walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -26,46 +26,106 @@
+    allBusyTimes.addAll(mandatoryTimes);
+    allBusyTimes.addAll(optionalTimes);
+    ArrayList<TimeRange> overlappingBusyTimes = getOverlap(allBusyTimes);
+    ArrayList<TimeRange> bothFree = findFreeTimes(overlappingBusyTimes, duration);
+    if (bothFree.isEmpty()) {
+      return findFreeTimes(getOverlap(mandatoryTimes), duration);
+    }
+    return bothFree;
+  }
+
+  /**
+  * Returns an ArrayList of TimeRanges of all of the times throughout the day
+  * that the given set of meeting attendees are busy and cannot be scheduled.
   *
   * @param  events the Collection of Event objects
-  * @param  requestAttendees the Set of attendees
-  * @return    the ArrayList of all meeting times that do not work with the attendees schedules
+  * @param  attendees the Set of meeting attendees
+  * @return   The ArrayList of TimeRanges where the given attendees are busy that day
   */
-  private ArrayList<TimeRange> getBusyTimes(Collection<Event> events, Set<String> requestAttendees) {
-    ArrayList<TimeRange> busyTimes = new ArrayList<TimeRange>();
-    for(Event event : events) {
+  private ArrayList<TimeRange> getBusyTimes(Collection<Event> events, Set<String> attendees) {
+    ArrayList<TimeRange> unavailableTimes = new ArrayList<TimeRange>();
+    for (Event event : events) {
       Set<String> eventAttendees = event.getAttendees();
-      Set<String> overlap = new HashSet<>(eventAttendees);
-      overlap.retainAll(requestAttendees);
-      if (!overlap.isEmpty()){
-        busyTimes.add(event.getWhen());
+      Set<String> attendeeOverlap = new HashSet<>(eventAttendees);
+      attendeeOverlap.retainAll(attendees);
+      if (!attendeeOverlap.isEmpty()) {
+        unavailableTimes.add(event.getWhen());
       }
     }
+    return unavailableTimes;
+  }
+
+  /**
+  * Returns an ArrayList of meeting times that are not availale
+  * and take into account overlapping meetings.
+  *
+  * @param  busyTimes an ArrayList that contains all of the unavailable meeting times
+  * @return    the ArrayList of all meeting times that don't work with overlapping schedules.
+  */
+  private ArrayList<TimeRange> getOverlap(ArrayList<TimeRange> busyTimes) {
     Collections.sort(busyTimes, TimeRange.ORDER_BY_START);
     ArrayList<TimeRange> busyWithOverlap = new ArrayList<TimeRange>();
-    if(busyTimes.size() > 0) {
+    if (busyTimes.size() > 0) {
       busyWithOverlap.add(busyTimes.get(0));
     }
     int overlapIndex = 0;
-    for (int busyIndex=0; busyIndex < busyTimes.size(); busyIndex++){
-      if (busyTimes.get(busyIndex).overlaps(busyWithOverlap.get(overlapIndex))){
+    for (int busyIndex=0; busyIndex < busyTimes.size(); busyIndex++) {
+      if (busyTimes.get(busyIndex).overlaps(busyWithOverlap.get(overlapIndex))) {
         int meetingStart = Math.min(busyWithOverlap.get(overlapIndex).start(), busyTimes.get(busyIndex).start());
         int meetingEnd = Math.max(busyWithOverlap.get(overlapIndex).end(), busyTimes.get(busyIndex).end()) - 1;
```

Comment:
Does this work if the second event ends before the first event ends?
Ex: if you have event 1 from 3:00 - 4:00, and event 2 from 3:15 - 3:30 -- will this say that the overlap is 3-4, or 3-3:30?

Reply:
I believe I've accounted for the case that you're bringing up here with the nested event test, if we use the example you provided it would have the total overlapping busy chunk be from 3-4!